libvast/vast/system/indexer_stage_driver.hpp

```diff
@@ -13,31 +13,73 @@
 
 #pragma once
 
-#include <caf/broadcast_downstream_manager.hpp>
-#include <caf/stream_stage_driver.hpp>
-
 #include "vast/fwd.hpp"
+#include "vast/system/circuit_downstream_manager.hpp"
 #include "vast/system/fwd.hpp"
+#include "vast/system/indexer.hpp"
 
-namespace vast::system {
+#include <caf/outbound_path.hpp>
+#include <caf/stream_stage_driver.hpp>
 
-/// @relates indexer_stage_driver
-/// Filter type for dispatching slices to INDEXER actors.
-using indexer_stage_filter = type;
+#include <cstddef>
+#include <unordered_map>
+#include <unordered_set>
+
+/// The indexer_stage_driver is responsible to receive a stream of table slices
+/// and demultiplex it into streams of table slice columns that are relayed to
+/// the to indexer actors in the target partition. This happens in 2 steps:
+/// First, the table id offset is used to determine the target partition.
+/// The partition is created if it did not exists beforehand. Second, the
+/// layout is used to retrieve the set of downstream slots that the slice shall
+/// be passed on to. If the table entry is missing, it is created by retrieving
+/// the indexer actors from the target partition and associating them to their
+/// matching fields from the layout.
+///
+///  Example for a partition containing 2 types foo and bar with the layouts:
+///
+///     field name: field type # ID (hash of record_field)
+///  type foo = record {
+///     a:          int,       # A
+///     b:          string,    # B
+///     c:          string,    # C
+///     d:          address    # D
+///  }
+///  type bar = record {
+///     a:          int,       # A
+///     b:          domain,    # B'
+///     c:          enum,      # C'
+///     d:          address    # D
+///  }
```

Comment:
I don't think you can map a field `a` that is in different layouts to the same indexer. What if `a` has completely different semantics in `foo` and `bar`?

Reply:
We changed the example so that both types are `foo`, but the second one has move precise field types.